### PR TITLE
CRUD API pgsql

### PR DESCRIPTION
**Pull request by samasno**
## Trello Ticket
https://trello.com/c/4wtYfizx/29-build-crud-api-for-postgres-and-write-api-and-db-tests

## Major additions
` database/postgres/data.js` - added data generators to use in middlewares for creating random docs in testing
`database/postgres/database.js` - postgres model for building raw queries and functions for crud api
`server/controllers/postgres.js` - crud api using postgres, includes middlewares for generating random test data
`server/controllers/utils.js` - custom utility functions for debugging, error handling and crud api

## Purpose of Pull Request
Creates a CRUD API to interact with newly seeded postgres database

### CODE REVIEW

File: database/postgres/database.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    info.brand,
+    info.isPrimeFreeOneDay,
+    info.isFreeDelivery,
+    similarItems.similarItems
+  from products
+  inner join configurations on products.configuration = configurations.id
+  inner join descriptions on products.description = descriptions.id
+  inner join info on products.info = info.id
+  inner join similarItems on products.similarItems = similarItems.id
+    ${Array.isArray(productIds) ? `
```
Comment (sjamcox) [inline database/postgres/database.js:25]: This is a clever way to customize the query based on input. Your comment says "get single product" though ... 🤔

File: database/postgres/database.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  INSERT INTO ${table}
+    (${columns.join(',')})
+  VALUES(${values.map(val => `'${val}'`).join(',')});
+`;
+
+// reusable delete query
+// where is prop=val tuple
+const deleteQuery = ({table = '', where = [['prop', 'val']]}) => `
+  DELETE FROM ${table}
+  WHERE ${where.map(tup => `${tup[0]}='${tup[1]}'`).join(' AND ')};
```
Comment (sjamcox) [inline database/postgres/database.js:67]: Multiple slick uses of .map()
Nice job.

File: database/postgres/database.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    products.similarItems
+  from products
+  where products.id = '${productId}';
+`;
+
+// build out query templates for CUD
+// reusable update table template
+// values passed as arrays must be stringified, where is pair of prop='val'
+// columns and values must be in same order
+const updateQuery = ({table = '', columns = [], values = [], where = [['prop', 'val']]}) => `
```
Comment (sjamcox) [inline database/postgres/database.js:47]: I like the reusable query structure you've set up.

File: server/controllers/postgres.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    req.body.productIds = productIds;
+  } 
+  next();
+});
+
+
+
+
+
```
Comment (sjamcox) [inline server/controllers/postgres.js:167]: The system you've put in place to handle transactions across multiple tables is pretty incredible 👍

**Review by sjamcox — APPROVED** [review]

File: database/postgres/database.js (hunk 1 above)
Comment (samasno) [reply database/postgres/database.js:25]: orignallly had it the single and batch in two seperate queries, forgot to update

**Pull request merged**